### ISSUE TITLE
Add Unit tests to the package mil.nga.giat.geowave.format.landsat8

### ISSUE THREAD

**Issue by rfecher** [ngageoint/geowave#999]
There are currently no unit tests of any of the public methods in the java classes inside the mil.nga.giat.geowave.format.landsat8 package.  These are commands exposed to the GeoWave command line tool that interact with the publicly available LandSat 8 dataset to make it easier for users to understand what’s available and ingest the data into GeoWave.

**Comment by mak438** [comment]
Should the unit tests be for the Command classes, or do the Iterator classes in this package require tests for their close(), next(), hasNext() methods as well?

**Comment by rfecher** [comment]
The *Runner classes do most of the work so I'd think they'd be prime candidates for tests, but also SceneFeatureIterator and WRS2GeometryStore are complex enough that they should have some testing.  The SceneFeatureIterator abstracts the concept of downloading the list of scenes from a public URL and filtering through the list based on user specified criteria, and the WRS2GeometryStore abstracts the concept of downloading the shapefile that contains all of the row/path pairs that represent the footprints that Landsat8 is collecting globally and iterates through the geometries based on some user-specified filter criteria.

**Comment by mak438** [comment]
Hi @rfecher. Thanks for the info. I began working on the tests for Landsat8AnalyzeCommand and Landsat8DownloadCommand (the *Runner classes don't have a public execute method and so I chose to look at the *Command classes instead).

I've noticed that both of these perform a large download and extract some files as part of their operation. I have some questions regarding the scope of these tests to make them most useful:

For the analyze command, I set up a test that calls Landsat8AnalyzeCommand#execute(), which downloads the files from the public URL. The test then verifies that certain values were printed to stdout ("Total Scenes: ", "Date Range: ", etc). Is there a way to mock the data source where this is fetched, perhaps by having a URL flag in Landsat8BasicCommandLineOptions? This would make the test go faster (since it doesn't need to download as much) and would let the test check that the computed totals are correct (I'm cautious to add that just yet in case data changes from http://landsat-pds.s3.amazonaws.com/ cause tests to fail).

The download command also loads some pretty large files (at least 150MB worth of images, could be more). Could that be mocked to load from a different source, perhaps something locally in a test folder, which would speed up that test as well? I haven't gotten this test completely working yet, so I'm not sure what else this entails.

In other words, is there a way (and would it be more useful) to have the tests be independent of the landsat data source, and simply validate that whatever's loading the data is working correctly?

Thanks in advance.

**Comment by rfecher** [comment]
the way to mock this up would probably be to use the `-ws` workspace directory and `--useCachedScenes`.  You could take a very small subset of the scenes listed in a mocked CSV and have it staged in that directory already. For now though, as a first step I think it may be simplest to focus on getting the tests working how we want it, and then the next step could be to re-use the cache in the landsat "workspace" directory to limit or eliminate external downloads (it uses this workspace concept as a cache, and the directory is configurable).  For example, I ran this command `landsat analyze --nbestscenes 1 --usecachedscenes --cql "BBOX(shape,-76.6,42.34,-76.4,42.54)"` and it didn't seem too bad - it looked liked the quality assessment band is 0.2 MB. So the following CQL filtter could be applied to a download command along with `--nbestscenes 1` that would limit the download to just that 0.2 MB file `"BBOX(shape,-76.6,42.34,-76.4,42.54) and band='BQA'"`

**Comment by mak438** [comment]
Getting this error when testing the ingest command? Is this something separate I need to install:

> WARNING: Failed to load the GDAL native libs. This is not a problem unless you need to use the GDAL plugins: they won't be enabled.
> java.lang.UnsatisfiedLinkError: no gdaljni in java.library.path

**Comment by rfecher** [comment]
to read the geotiffs that are downloaded you likely need GDAL. There are default readers, but I don't think they correctly read most if not all of those geotiffs (I at least remember having trouble on the 16-bit single-band images, don't recall if some of the bands are different).  You can install GDAL using this java utility on linux and windows (not sure where to get the equivalent native libraries for Mac OS X): https://github.com/ngageoint/geowave/blob/master/test/src/main/java/mil/nga/giat/geowave/test/InstallGdal.java#L19

Pass a single argument to put it in a specific directory - ultimately this will have to be automated by the tests so that continuous integration can ensure GDAL is installed prior to running any tests you write that depend on GDAL. We can move InstallGdal to a common location accessible by both modules such as geowave-adapter-raster.

**Comment by mak438** [comment]
I was able to get the InstallGdal working, however that gets loaded to ./target/temp/gdal by default. I tried adding that to the path from within the test, but ran into this [issue](http://fahdshariff.blogspot.jp/2011/08/changing-java-library-path-at-runtime.html). Can I set the directory where gdal gets installed to something like /usr/lib?

Trying to get this automated if I can.

**Comment by mak438** [comment]
Trying to run the test again (after copying gdal into /usr/lib), and I get the following issue:

```
OpenJDK 64-Bit Server VM warning: INFO: os::commit_memory(0x00000000ec41f000, 196608, 0) failed; error='Cannot allocate memory' (errno=12)
#
# There is insufficient memory for the Java Runtime Environment to continue.
# Native memory allocation (mmap) failed to map 196608 bytes for committing reserved memory.
# An error report file with more information is saved as:
# /home/moshe/Desktop/geowave_project/geowave/extensions/cli/landsat8/hs_err_pid2524.log
```

I'm trying to ingest into a `memory` store. Should the test create some other kind of store instead? What are the memory requirements to ingest from landsat8 -- I've given the JVM 1024MB already.

**Comment by rfecher** [comment]
ugh, that basically means the GDAL native library isn't setup correctly or is incompatible with your O/S

hs_er_pid's happen when native libraries fail, I don't think its a JVM memory issue at all (this memory is allocated outside of the JVM)

What O/S are you running?  I've tried it on several versions of Linux and on Windows but not on Mac OS X (or more precisely, I don't think it'd work on Mac OS X).

**Comment by mak438** [comment]
I'm running this in a VM, Ubuntu 16.04.1 LTS.

**Comment by rfecher** [comment]
hmm, thats what travis-ci is running on every commit...let me stub out something that takes care of the GDAL part automatically that will work on Ubuntu 16.04.1 LTS.

**Comment by mak438** [comment]
Thank you. I moved InstallGdal into geowave-adapter-raster, so it can be called by the test automatically. Do you want a pull request for that change?

**Comment by rfecher** [comment]
yes, please upen a PR when you have a chance that shows what you are currently doing with GDAL, at least it'll give me a chance to review what you are currently doing - perhaps there's a simple fix

The automated tests that are running on travis-ci use this to setup GDAL: https://github.com/ngageoint/geowave/blob/master/test/pom.xml#L216 (lines 216-274 in the pom, in particular notice where it sets the environment variables for LD_LIBRARY_PATH and PATH on line 243)

**Comment by mak438** [comment]
https://github.com/ngageoint/geowave/pull/1014

**Comment by rfecher** [comment]
The native libraries require the LD_LIBRARY_PATH and PATH to be setup similar to how the test/pom.xml linked above sets the environment variables.  If you did not manually set this yourself, this is the culprit for the issues you had.

**Comment by mak438** [comment]
FYI @rfecher, Adding environment variables from Eclipse -> Run Configurations -> Environment (I wasn't able to get it to work from maven) gives me the following error file.
[hs_err_pid4212.log.txt](https://github.com/ngageoint/geowave/files/755323/hs_err_pid4212.log.txt)

**Comment by rfecher** [comment]
I have GEOWAVE-999 branch with an initial test showing GDAL working through `mvn verify`: https://github.com/ngageoint/geowave/commit/77d0c2ede426159259b357f128ebf7434d4a4002

**Comment by rfecher** [comment]
keep in mind the key to getting it working through an IDE will either be to run `mvn verify` from the IDE or to ensure `LD_LIBRARY_PATH` and `PATH` point to the binaries downloaded and extracted from InstallGdal

**Comment by mak438** [comment]
Getting the following when trying to compile:

```
moshe@moshe-VirtualBox:~/Desktop/geowave_project/geowave/extensions/cli/landsat8$ pwd
/home/moshe/Desktop/geowave_project/geowave/extensions/cli/landsat8
moshe@moshe-VirtualBox:~/Desktop/geowave_project/geowave/extensions/cli/landsat8$ git log -n 1
commit c23a6b881a45a297970d0e84ab12942b9bc83437
Author: Rich Fecher <rfecher@gmail.com>
Date:   Tue Feb 7 12:01:58 2017 -0500

    inital test for raster ingest showing working GDAL
moshe@moshe-VirtualBox:~/Desktop/geowave_project/geowave/extensions/cli/landsat8$ git status
On branch GEOWAVE-999
Your branch is up-to-date with 'upstream/GEOWAVE-999'.
Untracked files:
  (use "git add <file>..." to include in what will be committed)

	src/test/java/mil/nga/giat/geowave/format/landsat8/Landsat8AnalyzeCommandTest.java
	src/test/java/mil/nga/giat/geowave/format/landsat8/Landsat8DownloadCommandTest.java
	src/test/java/mil/nga/giat/geowave/format/landsat8/Landsat8IngestCommandTest.java
	src/test/java/mil/nga/giat/geowave/format/landsat8/SceneFeatureIteratorTest.java
	src/test/java/mil/nga/giat/geowave/format/landsat8/WRS2GeometryStoreTest.java

nothing added to commit but untracked files present (use "git add" to track)
moshe@moshe-VirtualBox:~/Desktop/geowave_project/geowave/extensions/cli/landsat8$ mvn compile
[INFO] Scanning for projects...
[INFO]                                                                         
[INFO] ------------------------------------------------------------------------
[INFO] Building GeoWave LandSat8 Operations 0.9.4-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- jacoco-maven-plugin:0.7.4.201502262128:prepare-agent (pre-unit-test) @ geowave-cli-landsat8 ---
[INFO] jacoco.surefire.argline set to -javaagent:/home/moshe/.m2/repository/org/jacoco/org.jacoco.agent/0.7.4.201502262128/org.jacoco.agent-0.7.4.201502262128-runtime.jar=destfile=/home/moshe/Desktop/geowave_project/geowave/extensions/cli/landsat8/target/coverage-reports/jacoco-ut.exec,includes=mil/nga/giat/geowave/**
[INFO] 
[INFO] --- formatter-maven-plugin:0.5.2:format (default) @ geowave-cli-landsat8 ---
[INFO] Using 'UTF-8' encoding to format source files.
[INFO] Number of files to be formatted: 29
[INFO] Successfully formatted: 0 file(s)
[INFO] Fail to format        : 0 file(s)
[INFO] Skipped               : 29 file(s)
[INFO] Approximate time taken: 0s
[INFO] 
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ geowave-cli-landsat8 ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 2 resources
[INFO] 
[INFO] --- maven-compiler-plugin:2.3.2:compile (default-compile) @ geowave-cli-landsat8 ---
[INFO] Compiling 23 source files to /home/moshe/Desktop/geowave_project/geowave/extensions/cli/landsat8/target/classes
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /home/moshe/Desktop/geowave_project/geowave/extensions/cli/landsat8/src/main/java/mil/nga/giat/geowave/format/landsat8/RasterIngestRunner.java:[49,54] error: package mil.nga.giat.geowave.adapter.raster.plugin.gdal does not exist
[ERROR] /home/moshe/Desktop/geowave_project/geowave/extensions/cli/landsat8/src/main/java/mil/nga/giat/geowave/format/landsat8/RasterIngestRunner.java:[570,16] error: cannot find symbol
[ERROR]  class BandData
/home/moshe/Desktop/geowave_project/geowave/extensions/cli/landsat8/src/main/java/mil/nga/giat/geowave/format/landsat8/RasterIngestRunner.java:[576,10] error: cannot find symbol
[INFO] 3 errors 
[INFO] -------------------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 7.435 s
[INFO] Finished at: 2017-02-07T15:32:16-07:00
[INFO] Final Memory: 40M/124M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:2.3.2:compile (default-compile) on project geowave-cli-landsat8: Compilation failure: Compilation failure:
[ERROR] /home/moshe/Desktop/geowave_project/geowave/extensions/cli/landsat8/src/main/java/mil/nga/giat/geowave/format/landsat8/RasterIngestRunner.java:[49,54] error: package mil.nga.giat.geowave.adapter.raster.plugin.gdal does not exist
[ERROR] /home/moshe/Desktop/geowave_project/geowave/extensions/cli/landsat8/src/main/java/mil/nga/giat/geowave/format/landsat8/RasterIngestRunner.java:[570,16] error: cannot find symbol
[ERROR] class BandData
[ERROR] /home/moshe/Desktop/geowave_project/geowave/extensions/cli/landsat8/src/main/java/mil/nga/giat/geowave/format/landsat8/RasterIngestRunner.java:[576,10] error: cannot find symbol
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```

IDE doesn't show any problems.

**Comment by rfecher** [comment]
Try `mvn install`  from `/home/moshe/Desktop/geowave_project/geowave`.  The issue is you aren't getting the latest changes from geowave-adapter-raster when you are running mvn commands from the Landsat sub-module.  Installing the entire project to your local maven repository will fix that.

**Comment by mak438** [comment]
That worked, thank you.

**Comment by mak438** [comment]
RasterIngestRunnerTest has been running for over 25 minutes. Were you able to get the test to run in your environment?

`11065 moshe     20   0 3346916 1.580g   6572 S 32.3 80.9   3:51.97 java   `

**Comment by rfecher** [comment]
something's wrong if its taking 25 minutes - it ran on my local environment but more importantly ran on travis CI: https://travis-ci.org/ngageoint/geowave/jobs/199343293#L3664

It looks like the full suite of unit tests took 20 minutes on that run, whereas on master it took 18.5 minutes. So it may take about 1.5 minutes.  Did you run it from Eclipse previously?

**Comment by mak438** [comment]
I did, but that wasn't setting the environment variables correctly, so I tried it from maven next.

**Comment by mawhitby** [comment]
@mak438 Can you try running an mvn verify in eclipse? Just left click on the pom.xml file under geowave-cli-landsat8, under "Run As" select "2 Maven Build..." That will open a the Edit Configuration window. In the text box next to "Goals:" write: clean verify
Then hit run. This will build the geowave-cli-landsat8 project which will run your unit test and post the output to your console.

**Comment by mak438** [comment]
Same error as originally. I commented out the other test files, so it's just running this one. Ran with -X to get more output.

I do see the PATH and LD_LIBRARY_PATH are better now.
[hs_err_pid18271.log.txt](https://github.com/ngageoint/geowave/files/765061/hs_err_pid18271.log.txt)
[maven_command_output.txt](https://github.com/ngageoint/geowave/files/765062/maven_command_output.txt)
[contents_of_gdal_directory.txt](https://github.com/ngageoint/geowave/files/765060/contents_of_gdal_directory.txt)

**Comment by mawhitby** [comment]
Are you running this in a VM? If so make sure you are alocating at least 4GB to the machine itself. (That's the limit on TrivisCI and we know that it;s working there)
If you aren't running this on a VM and this is just your OS, we will have to go though it with you Saturday when we are up there. 

If this isn't working, try to focus on other parts of this package that don't require GDAL

**Comment by mak438** [comment]
I think I got it working now. Thanks for both your help.

**Comment by mawhitby** [comment]
Tests added in PR #1036
